presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -352,11 +346,45 @@ private Type getType(Expression expression)
             }
         }
 
-        private OptionalDouble doubleValueFromLiteral(Type type, Literal literal)
+        /**
+         * Returns expression value if the expression is constant, empty result otherwise.
+         */
+        private Optional<Object> tryEvaluateConstantExpression(Expression expression)
+        {
+            Map<NodeRef<Expression>, Type> expressionTypes = getExpressionTypes(expression);
+            ExpressionInterpreter interpreter = ExpressionInterpreter.expressionOptimizer(expression, metadata, session, expressionTypes);
+            Object value = interpreter.optimize(NoOpSymbolResolver.INSTANCE);
+
+            if (value instanceof Expression) {
+                // expression is not constant
+                return Optional.empty();
+            }
+            if (value == null) {
+                // expression evaluates to SQL null, which is unsupported here (expression should have been simplified not to be comparison expression earlier)
+                return Optional.empty();
+            }
+
+            return Optional.of(value);
+        }
+
+        private OptionalDouble doubleValueFromConstant(Type type, Object constantValue)
```

Comment:
`type` is used to construct the `DomainConverter`

> Also why leftType was used for right expression and vice-versa in previous version of this code?

This is still the case in my version too:   `tryEvaluateConstantExpression` returns only the value rather than `optional(pair(value, type))`.. we assume types of the comparison operands are the same.